addons/hr_expense/migrations/10.0.2.0/pre-migration.py

```diff
@@ -11,9 +12,50 @@
 }
 
 
+def _detect_hr_expense_v8(env):
+    """If you come from v8 and haven't cleaned up table, it's possible
+    to rescue previous expense sheets and use them in this version.
+    """
+    _table = "openupgrade_legacy_9_0_hr_expense_expense"
+    if not openupgrade.table_exists(env.cr, _table):
+        return
+    openupgrade.rename_models(
+        env.cr, [('hr.expense.expense', 'hr.expense.sheet')])
+    openupgrade.rename_tables(env.cr, [(_table, 'hr_expense_sheet')])
+    openupgrade.rename_fields(env, [
+        ('hr.expense.sheet', 'hr_expense_sheet', 'amount', 'total_amount'),
+        ('hr.expense.sheet', 'hr_expense_sheet', 'user_id', 'responsible_id'),
+        ('hr.expense', 'hr_expense', 'expense_id', 'sheet_id'),
+    ])
+    openupgrade.add_fields(env, [
+        # For avoiding to be filled by default value
+        ('bank_journal_id', 'hr.expense.sheet', 'hr_expense_sheet',
+         'many2one', False, 'hr_expense'),
+        ('accounting_date', 'hr.expense.sheet', 'hr_expense_sheet',
+         'date', False, 'hr_expense'),
+    ])
+    openupgrade.logged_query(
+        env.cr, """UPDATE hr_expense_sheet hes
+        SET accounting_date = am.date
+        FROM account_move am
+        WHERE am.id = hes.account_move_id""",
+    )
+    openupgrade.map_values(
+        env.cr, 'openupgrade_legacy_9_0_state', 'state', [
+            ('draft', 'submit'),
```

Comment:
maybe the ones in draft could be deleted, as it's not the same a confirmed (submited) one that a draft which was not even confirmed.

Reply:
Well, not really, but we don't want to lose those, and more having in v12 again the draft state. I have just pushed directly a little patch for it:

https://github.com/OCA/OpenUpgrade/commit/7abe19fa2220bb27b92e3d26901dfa762b018377

so I would go for this way.

Reply:
then don't delete them but don't change their state and left in draft although they are in an invalid state.

if not, imagine the following: coming from v8, with record in draft, then migrate to v10, and then stay in v10 production during some moths because someone doesn't to migrate too fast to v12. then this record in draft record in "submit" state follows some changes and and change the state to another state. then, few months later you migrate to v12 and then this record becomes again in 'draft' state! that doesn't makes sense!!!

And would you want to make a stop in v10 instead of migrating from v8 to v12 in one swoop? well, I could give you several reasons.

Reply:
Why they are in an invalid state? Being submitted doesn't mean they are anything more than. Semantically, if you go only from v8 to v10, you mustn't lose that expenses groups, but auto-submit them.

Reply:
ok, I see now that in the patch in v12 you only change to draft the ones that are still in submit (and were old draft).

Reply:
Yes, that's the safeguard part of the comment in the commit :wink: